src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,122 @@
+(defpackage #:coalton-impl/codegen/constant-propagation
+  (:use
+   #:cl
+   #:coalton-impl/codegen/ast)
+  (:local-nicknames
+   (#:tc #:coalton-impl/typechecker))
+  (:import-from
+   #:coalton-impl/codegen/traverse
+   #:action
+   #:*traverse*
+   #:traverse)
+  (:export
+   #:propagate-constants))
+
+(in-package #:coalton-impl/codegen/constant-propagation)
+
+(defun constant-var-value (var constant-bindings &key no-error)
+  (declare (type symbol var)
+           (values (or node null) &optional))
+  (let ((may-be-value (cdr (assoc var constant-bindings))))
+    (cond (may-be-value
+           ;; Whenever it exists, it is always a NODE (thus, non-NIL)
+           may-be-value)
+          (no-error
+           nil)
+          (t
+           (error "Expected VAR ~S to be a constant but is not" var)))))
+
+(defun constant-node-p (env node constant-bindings)
+  "If NODE is a constant, returns the constant value corresponding to it.
+If not, returns NIL"
+  (declare (type tc:environment env)
+           (type node node)
+           (type list constant-bindings))
+  (cond ((node-literal-p node)
+         (if (numberp (node-literal-value node))
```

Comment:
Why only numbers? The allowed types are:
```lisp
(deftype literal-value ()
  "Allowed literal values as Lisp objects."
  '(or integer ratio single-float double-float string character))
```

I guess if you are considering strings as not-necessarily-constant depending on whether we care about pointer equality, then it makes sense to not include strings by default. But shouldn't characters still be considered constant?

```suggestion
         (if (or (numberp (node-literal-value node)) (characterp (node-literal-value node)))
```

Reply:
It seems that even the case of numbers is contentious. ([Ref.](http://clhs.lisp.se/Body/f_eq.htm))

> The effect is that Common Lisp makes no guarantee that eq is true even when both its arguments are ``the same thing'' if that thing is a character or number.

Reply:
I don't have an opinion on this. I will be happy with whatever is suitable for coalton. If I recall, the main goal of constant propagation was to obtain the value of the `dict`-gensyms bound to codegen-symbols of the instances. And this has been achieved :D.

Reply:
I'd say anything that works with CL:EQL (excluding CL:EQ) are obvious candidates. Those would be:

- any CL number type
- characters 

(http://www.ai.mit.edu/projects/iiip/doc/CommonLISP/HyperSpec/Body/fun_eql.html)

Booleans work because of `cl:eq` but they're important enough for inlining etc. that they ought to be propagated.

Honestly though, I'm fine to punt on any cases.